plugin/fallback/fallback_test.go

```diff
@@ -0,0 +1,226 @@
+package fallback
+
+import (
+	"context"
+	"testing"
+
+	"github.com/miekg/dns"
+
+	"fmt"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/plugin/pkg/healthcheck"
+	"github.com/coredns/coredns/plugin/proxy"
+	"github.com/coredns/coredns/plugin/test"
+	"github.com/coredns/coredns/request"
+)
+
+// stubNextHandler is used to simulate a proxy plugin.
+// It returns a stub Handler that returns the rcode and err specified when invoked.
+func stubNextHandler(rcode int, err error) test.Handler {
+	return test.HandlerFunc(func(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+		w.WriteMsg(&dns.Msg{})
+		return rcode, err
+	})
+}
+
+// testProxyCreator implements the proxyCreator interface.
+// Used by unit test to verify that the proxyCreator.Create() method is called as expected
+type testProxyCreator struct {
+	expectedUpstream proxy.Upstream
+	called           int
+	t                *testing.T
+}
+
+func (c *testProxyCreator) Create(trace plugin.Handler, upstream proxy.Upstream) plugin.Handler {
+	c.called++
+
+	// Ensure that it is called with the expected Upstream
+	if c.expectedUpstream != nil && c.expectedUpstream != upstream {
+		c.t.Fatalf("Expected upstream passed to proxyCreator is '%v', but got '%v'",
+			c.expectedUpstream, upstream)
+	}
+
+	return &proxy.Proxy{Trace: trace, Upstreams: &[]proxy.Upstream{upstream}}
+}
+
+// dummyUpstream implements the proxy.Upstream interface
+// It is used to fake an upstream used for creating a proxy.
+type dummyUpstream struct {
+	rcode int
+}
+
+func (u dummyUpstream) From() string {
+	return ""
+}
+
+func (u dummyUpstream) Select() *healthcheck.UpstreamHost {
+	return nil
+}
+
+func (u dummyUpstream) IsAllowedDomain(s string) bool {
+	return false
+}
+
+func (u dummyUpstream) Exchanger() proxy.Exchanger {
+	return dummyExchanger{}
+}
+
+func (u dummyUpstream) Stop() error {
+	return nil
+}
+
+// dummyExchanger implements the proxy.Exchanger interface
+// It is used solely to implement the dummyUpstream above
+type dummyExchanger struct{}
+
+func (e dummyExchanger) Exchange(ctx context.Context, addr string, state request.Request) (*dns.Msg, error) {
+	return &dns.Msg{}, nil
+}
+
+func (e dummyExchanger) Protocol() string {
```

Comment:
These functions call all be put on 1 line

Reply:
Good idea. Will fix.